### PR TITLE
Safe casting and uint proto cleanup

### PR DESCRIPTION
**Pull request by riley-stride**
## What is the purpose of the change

1. Replace sf13/cast with our custom fork that checks for overflows
2. Throughout the codebase, use `cast.To__E` from our fork, rather than `cast.To___`. The `E` suffix indicates that we will return `out,err` rather than just `out`, so that we can catch errors! This makes the code more verbose, but safer. 
3. Change a bunch of int64 types to uint64s for extra safety and to reduce the number of casts we must perform

## Brief Changelog

1. Replace cast library in go.mod. Please review [the last few commits in our fork of the cast library](https://github.com/Stride-Labs/cast/commits/master).
2. Replace `cast.ToInt64()` => `cast.ToInt64E` and `cast.ToUint64` => `cast.ToUint64E` and catch any errors
3. Change the following proto struct attributes from uint64 to int64
- depositrecord epochNumber
- UserRedemptionRecord epochNumber
- depositrecord amount
- icaaccount balance
- MsgRedeemStake amt
- epochTracker nextEpochStartTime
- epochTracker duration
- hostZone stakedBal

## Testing and Verifying
unit and integration tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? n/a

## Notes
- we should refactor `AddDelegationToValidator` at some point to take in only uint amounts, and make a separate fn for `ReduceDelegationFromValidator`; currently, `AddDelegationToValidator` is used to both add and reduce, which is (a) confusing and (b) requires funky casting! IMO this is an issue for another PR since it's a non-trivial change

### CODE REVIEW

**Comment by asalzmann** [comment]
Can you create a pull request with the changes in the fork of the cast library?

File: x/stakeibc/keeper/hooks.go [hunk 1]
```diff
@@ -253,15 +279,19 @@ func (k Keeper) StakeExistingDepositsOnHostZones(ctx sdk.Context, epochNumber in
...
 
-func (k Keeper) TransferExistingDepositsToHostZones(ctx sdk.Context, epochNumber int64, depositRecords []recordstypes.DepositRecord) {
+func (k Keeper) TransferExistingDepositsToHostZones(ctx sdk.Context, epochNumber uint64, depositRecords []recordstypes.DepositRecord) {
 	transferDepositRecords := utils.FilterDepositRecords(depositRecords, func(record recordstypes.DepositRecord) (condition bool) {
 		return record.Status == recordstypes.DepositRecord_TRANSFER
 	})
-	ibcTimeoutBlocks := cast.ToUint64(k.GetParam(ctx, types.KeyIbcTimeoutBlocks))
+	ibcTimeoutBlocks, err := cast.ToUint64E(k.GetParam(ctx, types.KeyIbcTimeoutBlocks))
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("Could not get IBC timeout blocks: %s", err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/hooks.go:288]: %v I believe in the sprintf

File: x/stakeibc/keeper/hooks.go [hunk 2]
```diff
@@ -319,21 +349,31 @@ func (k Keeper) UpdateRedemptionRates(ctx sdk.Context, depositRecords []recordst
 
 		undelegatedBalance, error := k.GetUndelegatedBalance(zoneInfo, depositRecords)
 		if error != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Could not get undelegated balance for host zone %s: %s", zoneInfo.ChainId, error))
```
Comment (vish-stride) [inline x/stakeibc/keeper/hooks.go:352]: second %s should be %v

File: x/stakeibc/keeper/hooks.go [hunk 3]
```diff
@@ -319,21 +349,31 @@ func (k Keeper) UpdateRedemptionRates(ctx sdk.Context, depositRecords []recordst
...
 			return error
 		}
-		stakedBalance := zoneInfo.StakedBal
+		k.Logger(ctx).Info(fmt.Sprintf("undelegatedBalance: %d", undelegatedBalance))
+		stakedBalance, err := cast.ToInt64E(zoneInfo.GetStakedBal())
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Could not get staked balance for host zone %s: %s", zoneInfo.ChainId, err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/hooks.go:358]: second %s should be %v

File: x/stakeibc/keeper/hooks.go [hunk 4]
```diff
@@ -319,21 +349,31 @@ func (k Keeper) UpdateRedemptionRates(ctx sdk.Context, depositRecords []recordst
...
+			return err
+		}
+		k.Logger(ctx).Info(fmt.Sprintf("stakedBalance: %d", stakedBalance))
 		moduleAcctBalance, error := k.GetModuleAccountBalance(zoneInfo, depositRecords)
-
 		if error != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Could not get module account balance for host zone %s: %s", zoneInfo.ChainId, error))
```
Comment (vish-stride) [inline x/stakeibc/keeper/hooks.go:364]: second %s should be %v

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 5]
```diff
@@ -40,9 +40,15 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, "cannot unstake an amount g.t. staked balance on host zone: %d", msg.Amount)
 	}
 
+	amt, err := cast.ToInt64E(msg.Amount)
+	if err != nil {
+		k.Logger(ctx).Error("error casting RedeemStake msg.Amount to int64", "err", err)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:45]: fmt.Sprintf !

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 6]
```diff
@@ -40,9 +40,15 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
...
+		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, "invalid amount: %s", err)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:46]: fmt.Sprintf for the 2nd arg

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 7]
```diff
@@ -221,11 +221,18 @@ func (k Keeper) SubmitTxsEpoch(ctx sdk.Context, connectionId string, msgs []sdk.
 		return 0, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to get epoch tracker for %s", epochType)
 	}
 	// BUFFER by 5% of the epoch length
-	bufferSize := cast.ToInt64(k.GetParam(ctx, types.KeyBufferSize))
+	bufferSize, err := cast.ToUint64E(k.GetParam(ctx, types.KeyBufferSize))
+	if err != nil {
+		return 0, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to get buffer size: %s", err)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_submit_tx.go:226]: fmt.Sprintf on second arg

File: x/stakeibc/keeper/msg_server_liquid_stake.go [hunk 8]
```diff
@@ -75,7 +75,12 @@ func (k msgServer) LiquidStake(goCtx context.Context, msg *types.MsgLiquidStake)
 		k.Logger(ctx).Error("failed to find deposit record")
 		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no deposit record for epoch (%d)", strideEpochTracker.EpochNumber)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_liquid_stake.go:76]: fmt.Sprintf on second arg

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 9]
```diff
@@ -58,7 +64,7 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 	balance := k.bankKeeper.GetBalance(ctx, sender, coinDenom)
 	k.Logger(ctx).Info(fmt.Sprintf("Redemption issuer IBCDenom balance: %v%s", balance.Amount, balance.Denom))
 	k.Logger(ctx).Info(fmt.Sprintf("Redemption requested redemotion amount: %v%s", inCoin.Amount, inCoin.Denom))
-	if balance.Amount.LT(sdk.NewInt(msg.Amount)) {
+	if balance.Amount.LT(sdk.NewInt(amt)) {
 		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "balance is lower than redemption amount. redemption amount: %d, balance %d: ", msg.Amount, balance.Amount)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:68]: fmt.Sprintf on second arg

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 10]
```diff
@@ -221,11 +221,18 @@ func (k Keeper) SubmitTxsEpoch(ctx sdk.Context, connectionId string, msgs []sdk.
...
+	}
 	BUFFER := epochTracker.Duration / bufferSize
 	timeoutNanos := epochTracker.NextEpochStartTime - BUFFER
-	k.Logger(ctx).Info(fmt.Sprintf("Submitting txs for epoch %s %d %d", epochTracker.EpochIdentifier, epochTracker.NextEpochStartTime, timeoutNanos))
-	sequence, err := k.SubmitTxs(ctx, connectionId, msgs, account, cast.ToUint64(timeoutNanos), callbackId, callbackArgs)
+	timeoutNanosSafe, err := cast.ToUint64E(timeoutNanos)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to get timeout nanos: %s", err)
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_submit_tx.go:232]: sdkerrors.Wrapf(sdkerrors

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 11]
```diff
@@ -342,7 +348,11 @@ func (k Keeper) GetLightClientHeightSafely(ctx sdk.Context, connectionID string)
 	} else {
 		// TODO(TEST-119) get stAsset supply at SAME time as hostZone height
 		// TODO(TEST-112) check on safety of castng uint64 to int64
-		latestHeightHostZone = cast.ToInt64(clientState.GetLatestHeight().GetRevisionHeight())
+		latestHeightHostZone, err := cast.ToUint64E(clientState.GetLatestHeight().GetRevisionHeight())
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("error casting latest height to int64: %s", err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/msg_server_submit_tx.go:353]: %v

File: x/stakeibc/keeper/unbonding_records.go [hunk 12]
```diff
@@ -71,8 +71,16 @@ func (k Keeper) SendHostZoneUnbondings(ctx sdk.Context, hostZone types.HostZone)
 	overflowAmt := int64(0)
 	for _, validator := range validators {
 		valAddr := validator.GetAddress()
-		valUnbondAmt := cast.ToInt64(newUnbondingToValidator[valAddr])
-		currentAmtStaked := cast.ToInt64(validator.GetDelegationAmt())
+		valUnbondAmt, err := cast.ToInt64E(newUnbondingToValidator[valAddr])
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Error casting validator unbonding amount %d: %s", newUnbondingToValidator[valAddr], err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/unbonding_records.go:76]: %v on err

File: x/stakeibc/keeper/unbonding_records.go [hunk 13]
```diff
@@ -71,8 +71,16 @@ func (k Keeper) SendHostZoneUnbondings(ctx sdk.Context, hostZone types.HostZone)
...
+			return false
+		}
+		currentAmtStaked, err := cast.ToInt64E(validator.GetDelegationAmt())
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Error casting validator staked amount %d: %s", validator.GetDelegationAmt(), err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/unbonding_records.go:81]: %v on err

File: x/stakeibc/keeper/unbonding_records.go [hunk 14]
```diff
@@ -85,7 +93,12 @@ func (k Keeper) SendHostZoneUnbondings(ctx sdk.Context, hostZone types.HostZone)
 			valUnbondAmt := valAddrToUnbondAmt[valAddr]
 			currentAmtStaked := validator.GetDelegationAmt()
 			// store how many more tokens we could unbond, if needed
-			amtToPotentiallyUnbond := cast.ToInt64(currentAmtStaked) - valUnbondAmt
+			curAmtStaked, err := cast.ToInt64E(currentAmtStaked)
+			if err != nil {
+				k.Logger(ctx).Error(fmt.Sprintf("Error casting validator staked amount %d: %s", currentAmtStaked, err))
```
Comment (vish-stride) [inline x/stakeibc/keeper/unbonding_records.go:98]: %v on err

File: x/stakeibc/keeper/unbonding_records.go [hunk 15]
```diff
@@ -229,7 +242,12 @@ func (k Keeper) SweepAllUnbondedTokens(ctx sdk.Context) {
 				delegationAccount := zoneInfo.GetDelegationAccount()
 				redemptionAccount := zoneInfo.GetRedemptionAccount()
 
-				sweepCoin := sdk.NewCoin(zoneInfo.HostDenom, sdk.NewInt(cast.ToInt64(totalAmtTransferToRedemptionAcct)))
+				totalAmtTransferToRedemptionAcct, err := cast.ToInt64E(totalAmtTransferToRedemptionAcct)
+				if err != nil {
+					k.Logger(ctx).Error(fmt.Sprintf("\t\tCould not convert %d to int64", totalAmtTransferToRedemptionAcct))
+					return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "\t\tCould not convert %d to int64", totalAmtTransferToRedemptionAcct)
```
Comment (vish-stride) [inline x/stakeibc/keeper/unbonding_records.go:248]: fmt.sprintf

**Review by vish-stride — CHANGES_REQUESTED** [review]
Looks great, really glad we're adding these safety checks!

Some minor formatting (e.g. sprintf) things, but otherwise looks wonderful

File: x/stakeibc/keeper/unbonding_records.go [hunk 16]
```diff
@@ -85,7 +93,12 @@ func (k Keeper) SendHostZoneUnbondings(ctx sdk.Context, hostZone types.HostZone)
 			valUnbondAmt := valAddrToUnbondAmt[valAddr]
 			currentAmtStaked := validator.GetDelegationAmt()
 			// store how many more tokens we could unbond, if needed
-			amtToPotentiallyUnbond := cast.ToInt64(currentAmtStaked) - valUnbondAmt
+			curAmtStaked, err := cast.ToInt64E(currentAmtStaked)
```
Comment (asalzmann) [inline x/stakeibc/keeper/unbonding_records.go:96]: why is this being recast to an int64?

**Review by asalzmann — APPROVED** [review]
lgtm - +1 to using Sprintf and a few small comments, but once those are addressed good to merge!

File: x/stakeibc/keeper/host_zone.go [hunk 17]
```diff
@@ -100,10 +100,19 @@ func (k Keeper) AddDelegationToValidator(ctx sdk.Context, hostZone types.HostZon
...
-				val.DelegationAmt = val.GetDelegationAmt() + cast.ToUint64(amt)
+				amt, err := cast.ToUint64E(amt)
+				if err != nil {
+					k.Logger(ctx).Error(fmt.Sprintf("Error converting %d to uint64", amt))
+					return false
+				}
+				val.DelegationAmt = val.GetDelegationAmt() + amt
 				return true
 			} else {
-				absAmt := cast.ToUint64(-amt)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/host_zone.go:106]: 100% agree as discussed offline but let's do it in a new PR to keep this one to type changes, will add to JIRA

File: x/stakeibc/keeper/unbonding_records.go (hunk 16 above)
Comment (riley-stride) [reply x/stakeibc/keeper/unbonding_records.go:96]: fixed

File: x/stakeibc/keeper/validator_selection.go [hunk 18]
```diff
@@ -59,7 +69,11 @@ func (k Keeper) GetTargetValAmtsForHostZone(ctx sdk.Context, hostZone types.Host
 			// for the last element, we need to make sure that the allocatedAmt is equal to the finalDelegation
 			targetAmount[validator.GetAddress()] = finalDelegation - allocatedAmt
 		} else {
-			delegateAmt := cast.ToUint64(float64(validator.Weight*finalDelegation) / float64(totalWeight))
+			delegateAmt, err := cast.ToUint64E(float64(validator.Weight*finalDelegation) / float64(totalWeight))
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/validator_selection.go:72]: we should prob use sdk.dec instead here for consistency across the codebase, wdyt @vishal?

File: x/interchainquery/keeper/keeper.go [hunk 19]
```diff
@@ -107,7 +107,11 @@ func (k *Keeper) GetDatapointOrRequest(ctx sdk.Context, module string, connectio
 		return types.DataPoint{}, fmt.Errorf("no data; query submitted")
 	}
 
-	if val.LocalHeight.LT(sdk.NewInt(ctx.BlockHeight() - cast.ToInt64(max_age))) { // this is somewhat arbitrary; TODO: make this better
+	max_age_, err := cast.ToInt64E(max_age)
+	if err != nil {
+		return types.DataPoint{}, err
+	}
+	if val.LocalHeight.LT(sdk.NewInt(ctx.BlockHeight() - max_age_)) { // this is somewhat arbitrary; TODO: make this better
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/interchainquery/keeper/keeper.go:114]: afaiu it's a comment about unfinished ICQ ttl implementation, has been here from the start

**Comment by riley-stride** [comment]
@vish-stride can you approve this? you requested a change, I've incorporated them all but it's blocking me until you re-approve

File: proto/records/genesis.proto [hunk 20]
```diff
@@ -16,7 +16,7 @@ message UserRedemptionRecord {
   uint64 amount = 4; 
   string denom = 5; 
   string hostZoneId = 6; 
-  int64 epochNumber = 7; 
+  uint64 epochNumber = 7;
```
[the remark this answers is not in the document]
Comment (vish-stride) [reply proto/records/genesis.proto:19]: Normally yes, but I believe the way we serialize/deserialize in the genesis file is by casting to string, so we should be okay here. Later today I’ll get our intended genesis file and we can test this out [](edited-with-Axolo)

File: x/stakeibc/keeper/validator_selection.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (vish-stride) [reply x/stakeibc/keeper/validator_selection.go:72]: Yeah agreed, sdk.dec is better I think. floats are also stochastic so have a high chance of causing a consensus failure later [](edited-with-Axolo)

**Review by vish-stride — APPROVED** [review]
Looks wonderful

**Pull request merged**